app/src/main/java/fr/gaulupeau/apps/Poche/data/QueueHelper.java

```diff
@@ -1,307 +1,325 @@
 package fr.gaulupeau.apps.Poche.data;
 
-import android.text.TextUtils;
 import android.util.Log;
 
-import java.util.Arrays;
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
 import java.util.EnumSet;
+import java.util.HashMap;
 import java.util.HashSet;
-import java.util.LinkedList;
 import java.util.List;
+import java.util.Map;
 import java.util.Set;
 
 import fr.gaulupeau.apps.Poche.data.dao.DaoSession;
 import fr.gaulupeau.apps.Poche.data.dao.QueueItemDao;
+import fr.gaulupeau.apps.Poche.data.dao.entities.AddLinkItem;
+import fr.gaulupeau.apps.Poche.data.dao.entities.AddOrUpdateAnnotationItem;
+import fr.gaulupeau.apps.Poche.data.dao.entities.ArticleChangeItem;
+import fr.gaulupeau.apps.Poche.data.dao.entities.ArticleDeleteItem;
+import fr.gaulupeau.apps.Poche.data.dao.entities.ArticleTagsDeleteItem;
+import fr.gaulupeau.apps.Poche.data.dao.entities.DeleteAnnotationItem;
 import fr.gaulupeau.apps.Poche.data.dao.entities.QueueItem;
+import fr.gaulupeau.apps.Poche.data.dao.entities.SpecificItem;
 
 import static fr.gaulupeau.apps.Poche.data.dao.entities.QueueItem.*;
 
 public class QueueHelper {
 
-    private static final String TAG = QueueHelper.class.getSimpleName();
+    public static class QueueState {
 
-    private final QueueItemDao queueItemDao;
+        private List<QueueItem> queueItems;
 
-    public QueueHelper(DaoSession daoSession) {
-        queueItemDao = daoSession.getQueueItemDao();
-    }
+        private Set<QueueItem> negatedItems;
```

Comment:
These are the items whose actions don't need to be performed because of some more recent items. For example, if a user updates article title and then removes the same article, the title update item will be negated - only deletion will be performed. The `negatedByDeletion` set contains the action types that may be negated that way.

Also, consecutive article changes are "collapsed" into one: the last item accumulates all the changes (to be sent at once) and all the previous ones are negated.